### PR TITLE
Fix Out of memory with source gen tests

### LINKED ISSUE #87934: Test failure: Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage
Also, Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial and others also fail, depending on the run.

All on Windows/x86. JitStress is not required. It does seem to require a Checked runtime, and does not fail with a Release runtime.

E.g.,

https://dev.azure.com/dnceng-public/public/_build/results?buildId=317311&view=ms.vss-test-web.build-test-results-tab&runId=6505130&resultId=137246&paneView=debug

```
  Starting:    Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests (parallel test collections = on, max threads = 4)
    Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial [FAIL]
      System.OutOfMemoryException : Insufficient memory to continue the execution of the program.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs(31,0): at System.Runtime.InteropServices.Marshal.AllocHGlobal(Int32 cb)
        /_/src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/NativeHeapMemoryBlock.cs(32,0): at System.Reflection.Internal.NativeHeapMemoryBlock.DisposableData..ctor(Int32 size)
        /_/src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/NativeHeapMemoryBlock.cs(62,0): at System.Reflection.Internal.NativeHeapMemoryBlock..ctor(Int32 size)
        /_/src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/StreamMemoryBlockProvider.cs(71,0): at System.Reflection.Internal.StreamMemoryBlockProvider.ReadMemoryBlockNoLock(Stream stream, Int64 start, Int32 size)
        /_/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.cs(190,0): at System.Reflection.PortableExecutable.PEReader..ctor(Stream peStream, PEStreamOptions options, Int32 size)
        /_/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.cs(128,0): at System.Reflection.PortableExecutable.PEReader..ctor(Stream peStream, PEStreamOptions options)
           at Microsoft.CodeAnalysis.ModuleMetadata.CreateFromStream(Stream peStream, PEStreamOptions options)
        /_/src/libraries/Common/tests/SourceGenerators/RoslynTestUtils.cs(38,0): at SourceGenerators.Tests.RoslynTestUtils.CreateTestProject(IEnumerable`1 references, Boolean includeBaseReferences, LanguageVersion langVersion)
        /_/src/libraries/Common/tests/SourceGenerators/RoslynTestUtils.cs(159,0): at SourceGenerators.Tests.RoslynTestUtils.RunGenerator(IIncrementalGenerator generator, IEnumerable`1 references, IEnumerable`1 sources, Boolean includeBaseReferences, LanguageVersion langVersion, CancellationToken cancellationToken)
        /_/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorParserTests.cs(947,0): at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.RunGenerator(String code, Boolean wrap, Boolean inNamespace, Boolean includeBaseReferences, Boolean includeLoggingReferences, CancellationToken cancellationToken)
        /_/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorParserTests.cs(609,0): at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial()
        --- End of stack trace from previous location ---
```

It's not clear when this started happening. The last `runtime-coreclr libraries-jitstress` pipeline without this failure was https://dev.azure.com/dnceng-public/public/_build/results?buildId=316592&view=results. However, I can repro the failure locally even with this git hash.

```json
{
  "ErrorMessage": "Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage [FAIL]",
  "BuildRetry": false,
  "ErrorPattern": "",
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=317311
**Result validation: :white_check_mark:** Known issue matched with the provided build.

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[326048](https://dev.azure.com/dnceng-public/public/_build/results?buildId=326048)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=326048&view=ms.vss-test-web.build-test-results-tab&runId=6709076&resultId=126971)||
|[326002](https://dev.azure.com/dnceng-public/public/_build/results?buildId=326002)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=326002&view=ms.vss-test-web.build-test-results-tab&runId=6708030&resultId=144840)||
|[325506](https://dev.azure.com/dnceng-public/public/_build/results?buildId=325506)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=325506&view=ms.vss-test-web.build-test-results-tab&runId=6694044&resultId=137040)||
|[325428](https://dev.azure.com/dnceng-public/public/_build/results?buildId=325428)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.Valid_AdditionalAttributes](https://dev.azure.com/dnceng-public/public/_build/results?buildId=325428&view=ms.vss-test-web.build-test-results-tab&runId=6691982&resultId=104225)||
|[324865](https://dev.azure.com/dnceng-public/public/_build/results?buildId=324865)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=324865&view=ms.vss-test-web.build-test-results-tab&runId=6676924&resultId=114314)||
|[324820](https://dev.azure.com/dnceng-public/public/_build/results?buildId=324820)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=324820&view=ms.vss-test-web.build-test-results-tab&runId=6676038&resultId=145732)||
|[323832](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323832)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323832&view=ms.vss-test-web.build-test-results-tab&runId=6665066&resultId=137263)|dotnet/runtime#88130|
|[324303](https://dev.azure.com/dnceng-public/public/_build/results?buildId=324303)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=324303&view=ms.vss-test-web.build-test-results-tab&runId=6664838&resultId=104112)||
|[323471](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323471)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323471&view=ms.vss-test-web.build-test-results-tab&runId=6645710&resultId=112488)||
|[323384](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323384)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323384&view=ms.vss-test-web.build-test-results-tab&runId=6642970&resultId=143397)||
|[323067](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323067)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=323067&view=ms.vss-test-web.build-test-results-tab&runId=6640620&resultId=152646)|dotnet/runtime#88090|
|[322854](https://dev.azure.com/dnceng-public/public/_build/results?buildId=322854)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=322854&view=ms.vss-test-web.build-test-results-tab&runId=6629462&resultId=134448)||
|[321944](https://dev.azure.com/dnceng-public/public/_build/results?buildId=321944)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=321944&view=ms.vss-test-web.build-test-results-tab&runId=6607788&resultId=124094)||
|[321895](https://dev.azure.com/dnceng-public/public/_build/results?buildId=321895)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=321895&view=ms.vss-test-web.build-test-results-tab&runId=6606604&resultId=141656)||
|[321311](https://dev.azure.com/dnceng-public/public/_build/results?buildId=321311)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=321311&view=ms.vss-test-web.build-test-results-tab&runId=6591890&resultId=137768)||
|[320380](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320380)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320380&view=ms.vss-test-web.build-test-results-tab&runId=6571414&resultId=111936)||
|[320333](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320333)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320333&view=ms.vss-test-web.build-test-results-tab&runId=6568500&resultId=144463)||
|[320129](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320129)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320129&view=ms.vss-test-web.build-test-results-tab&runId=6563750&resultId=141365)||
|[320072](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320072)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=320072&view=ms.vss-test-web.build-test-results-tab&runId=6562656&resultId=144979)||
|[319899](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319899)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319899&view=ms.vss-test-web.build-test-results-tab&runId=6558976&resultId=104214)||
|[319891](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319891)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319891&view=ms.vss-test-web.build-test-results-tab&runId=6558604&resultId=142227)||
|[319730](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319730)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319730&view=ms.vss-test-web.build-test-results-tab&runId=6554796&resultId=139822)|dotnet/runtime#87917|
|[319710](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319710)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.Valid_AdditionalAttributes](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319710&view=ms.vss-test-web.build-test-results-tab&runId=6554150&resultId=113153)||
|[319629](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319629)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319629&view=ms.vss-test-web.build-test-results-tab&runId=6552726&resultId=146315)||
|[319175](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319175)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319175&view=ms.vss-test-web.build-test-results-tab&runId=6546688&resultId=138870)|dotnet/runtime#87967|
|[319201](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319201)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=319201&view=ms.vss-test-web.build-test-results-tab&runId=6543112&resultId=142414)|dotnet/runtime#87917|
|[318922](https://dev.azure.com/dnceng-public/public/_build/results?buildId=318922)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=318922&view=ms.vss-test-web.build-test-results-tab&runId=6537060&resultId=120680)||
|[318763](https://dev.azure.com/dnceng-public/public/_build/results?buildId=318763)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=318763&view=ms.vss-test-web.build-test-results-tab&runId=6533752&resultId=141473)||
|[318753](https://dev.azure.com/dnceng-public/public/_build/results?buildId=318753)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NeedlessExceptionInMessage](https://dev.azure.com/dnceng-public/public/_build/results?buildId=318753&view=ms.vss-test-web.build-test-results-tab&runId=6533718&resultId=140007)||
|[317311](https://dev.azure.com/dnceng-public/public/_build/results?buildId=317311)|dotnet/runtime|[Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorParserTests.NotPartial](https://dev.azure.com/dnceng-public/public/_build/results?buildId=317311&view=ms.vss-test-web.build-test-results-tab&runId=6505130&resultId=137245)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|4|29|30|

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/87934

### CODE REVIEW

**Comment by BruceForstall** [comment]
Related: https://github.com/dotnet/runtime/issues/87934

**Comment by tarekgh** [comment]
@CyrusNajmabadi could you please have a look? 

CC @jaredpar

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by tarekgh** [comment]
The unrelated failure in the jitstress is tracked by https://github.com/dotnet/runtime/issues/88224

File: src/libraries/Common/tests/SourceGenerators/RoslynTestUtils.cs [hunk 1]
```diff
@@ -215,7 +228,8 @@ public static TextSpan MakeSpan(string text, int spanNum)
             string? defaultNamespace = null,
             string? extraFile = null)
         {
-            Project proj = CreateTestProject(references);
+            using var workspace = CreateTestWorkspace();
```
Comment (eiriktsarpalis) [inline src/libraries/Common/tests/SourceGenerators/RoslynTestUtils.cs:231]: For my education, OOM's were triggered by `AdhocWorkspace` instances not being disposed?

Comment (tarekgh) [reply src/libraries/Common/tests/SourceGenerators/RoslynTestUtils.cs:231]: that is right

**Pull request merged**